drivers/arm/ws2812.c

```diff
@@ -0,0 +1,111 @@
+#include "quantum.h"
+#include "ws2812.h"
+#include "ch.h"
+#include "hal.h"
+
+#if !defined(LED_ARRAY) && defined(RGB_MATRIX_ENABLE)
+// LED color buffer
+LED_TYPE led[DRIVER_LED_TOTAL];
+#    define LED_ARRAY led
+#endif
+
+#ifdef RGB_MATRIX_ENABLE
+// Set an led in the buffer to a color
+void inline ws2812_setled(int i, uint8_t r, uint8_t g, uint8_t b) {
+    led[i].r = r;
+    led[i].g = g;
+    led[i].b = b;
+}
+
+void ws2812_setled_all(uint8_t r, uint8_t g, uint8_t b) {
+    for (int i = 0; i < sizeof(led) / sizeof(led[0]); i++) {
+        led[i].r = r;
+        led[i].g = g;
+        led[i].b = b;
+    }
+}
+#endif
+
+/* adapted from
+https://github.com/bigjosh/SimpleNeoPixelDemo/blob/master/SimpleNeopixelDemo/SimpleNeopixelDemo.ino
+*/
+
+#define NS_PER_SEC (1000000000L)  // Note that this has to be SIGNED since we want to be able to check for negative values of derivatives
+
+#define CYCLES_PER_INSTRUCTION 0.35
+
+#define CYCLES_PER_SEC (STM32_SYSCLK * CYCLES_PER_INSTRUCTION)
+
+#define NS_PER_CYCLE (NS_PER_SEC / CYCLES_PER_SEC)
```

Comment:
As is probably already planned, need to allow overrides via `#ifndef`

Reply:
`CYCLES_PER_INSTRUCTION` is what i had imagined would be the fudge factor per chip, the latest changes being the process of setting up defaults but allowing user config.